spec/component/models/contributor_name_merge_policy_spec.rb

```diff
@@ -0,0 +1,150 @@
+            before do
+              contributor1.update user_id: 1
+            end
+
+            it 'selects that contributor as the preferred contributor' do
+              expect(policy.contributor_names_to_keep.first).to eq contributor1
+            end
+          end
+
+          context 'when both contributors have a user_id but one also has a role' do
+            before do
+              contributor1.update user_id: 1
+              contributor2.update user_id: 1
+              contributor1.update role: 'Author'
+            end
+
+            it 'selects that contributor as the preferred contributor' do
+              expect(policy.contributor_names_to_keep.first).to eq contributor1
+            end
+          end
+
+          context 'when one contributor has a role and the other does not' do
+            before do
+              contributor1.update role: 'Author'
+            end
+
+            it 'selects that contributor as the preferred contributor' do
+              expect(policy.contributor_names_to_keep.first).to eq contributor1
+            end
+          end
+
+          context 'when both contributors have a role but one also has a user_id' do
+            before do
+              contributor1.update role: 'Author'
+              contributor2.update role: 'Author'
+              contributor1.update user_id: 1
+            end
+
+            it 'selects that contributor as the preferred contributor' do
+              expect(policy.contributor_names_to_keep.first).to eq contributor1
+            end
+          end
+
+          context 'when both contributors have a user_id, and role' do
+            before do
+              contributor1.update user_id: 1
+              contributor2.update user_id: 1
+              contributor1.update role: 'Author'
+              contributor2.update role: 'Author'
+            end
+
+            it 'selects the contributor with a longer name as the preferred contributor' do
+              expect(policy.contributor_names_to_keep.first).to eq contributor1
+            end
+          end
+        end
+
+        context 'when contributors have different positions' do
+          let(:contributor1) { ContributorName.create first_name: 'Test', last_name: 'Tester', position: 1 }
+          let(:contributor2) { ContributorName.create first_name: 'Terry', last_name: 'Tester', position: 2 }
+
+          it 'treats these contributors as the different' do
+            expect(policy.contributor_names_to_keep.count).to eq 2
+          end
+        end
+      end
+
+      context 'when contributors have a different last name' do
+        context 'when contributors have the same position' do
+          let(:contributor1) { ContributorName.create first_name: 'Test', last_name: 'Tester', position: 1 }
+          let(:contributor2) { ContributorName.create first_name: 'Test', last_name: 'Terry', position: 1 }
+
+          it 'treats these contributors as the different' do
```

Comment:
@ajkiessl I'll post more comments, but here's one observation:

I've seen real-world examples where I think this behavior wouldn't be what we want. We sometimes think of a person's name as an immutable (if not unique) identifier for the person, but unfortunately it's not. I've seen cases where it looked like a person was listed as a contributor under their original last name in one set of publication metadata and under the last name that they took when they got married in a duplicate set of the publication metadata. Same person, same position, but different last name. Ideally we'd only be getting metadata that lists peoples' names as they were actually printed in the publication text, but I know that this isn't always the case.

So it looks like we'd end up listing two different names for the same person in cases like that. Does that sound right to you?

While it's less common, people do change their first names for various reasons too. So the same thing could conceivably happen with the first name comparison, though I haven't seen an actual example of this yet.

This is one reason why I started to tie ContributorName records to User records whenever possible. That could allow us to look up a User by ContributorName metadata and see any other names under which they may have published research. That could help us to determine whether or not two contributors with different names are actually the same person. Unfortunately, it can be a chicken and egg problem in terms of getting that data to begin with. At some point, someone has to tell us that it's actually the same person with different names.

So unfortunately I was never able to come up with a decent solution for this problem..

Reply:
@EricDurante That sounds right to me.  I'll add something to the ContributorNameMergePolicy that keeps unique contributor names on their `user_id` (if present) after all the filtering.  That way if we have Jane Smith and Jane Johnson that are both linked to the same `User` record, we'd know these are the same person and can get rid of one.  That could help fix some of the instances above.

I'm not sure there's any way right now to handle cases when we don't have a `user_id`, though.

Reply:
Yeah, there may not be. Perhaps there will be few enough instances of this that it won't be a show-stopper. Probably the worst case scenario at this point is that people who are using our API to populate websites with researcher/publication data will end up displaying one of these mistakes every now and then until someone notices it and we can correct it. It won't look great, but it might be pretty rare.

In the case where we do have a Jane Smith and a Jane Johnson who turn out to be the same person and we need to decide which name to keep, we might just have to go with the more reliable data source (if there is one).